DEBUGINFOD based DWP acquisition for LLDB

kevinfrei opened the pull request:
I've plumbed the LLVM DebugInfoD client into LLDB, and added automatic downloading of DWP files to the SymbolFileDWARF.cpp plugin. If you have DEBUGINFOD_URLS set to a space delimited set of web servers, LLDB will try to use them as a last resort when searching for DWP files. If you do *not* have that environment variable set, nothing should be changed. There's also a setting, per @clayborg 's suggestion, that will override the environment variable, or can be used instead of the environment variable. The setting is why I also needed to add an API to the llvm-debuginfod library

### Test Plan:

Suggestions are welcome here. I should probably have some positive and negative tests, but I wanted to get the diff up for people who have a clue what they're doing to rip it to pieces before spending too much time validating the initial implementation.

Commit:
DEBUGINFOD based DWP acquisition for LLDB

Summary:
I've plumbed the LLVM DebugInfoD client into LLDB, and added automatic downloading of
DWP files to the SymbolFileDWARF.cpp plugin. If you have `DEBUGINFOD_URLS` set to a
space delimited set of web servers, LLDB will try to use them as a last resort when
searching for DWP files. If you do *not* have that environment variable set, nothing
should be changed. There's also a setting, per Greg Clayton's request, that will
override the env variable, or can be used instead of the env var. This setting is the
reason for the additional API added to the llvm's Debuginfod library.

Test Plan:
Suggestions are welcome here. I should probably have some positive and negative tests,
but I wanted to get the diff up for people who have a clue what they're doing to rip it
to pieces before spending too much time validating my implementation.

walter-erquinigo commented on lldb/source/Symbol/LocateSymbolFile.cpp:
```diff
@@ -396,8 +398,22 @@ Symbols::LocateExecutableSymbolFile(const ModuleSpec &module_spec,
+    if (result)
+      return FileSpec(*result);
+    // An error is just fine, here...
+    consumeError(result.takeError());
```
log this to some log channel instead of consuming it

kevinfrei replied:
That was my reason for just consuming it in the first place. I think adding a "debuginfod_failure_log" setting makes sense here. And I just signed myself up for more work :/

mysterymath replied:
Looks like gdb has a `set debuginfod verbose on/off` setting for this. There's also the generic `DEBUGINFOD_VERBOSE` environment variable. If set, this triggers remote logging to stderr.

The environment variable is something we generally want to support for debuginfod lookups; that could probably safely live in the debuginfod client library itself, as it does with libdebuginfod. If LLDB had its own setting for this, it may be able to invoke a setter in the client library, but current everything in the client library is global, so there may be some threading gotchas there.

For reference, here's the gdb settings for debuginfod: https://sourceware.org/gdb/current/onlinedocs/gdb.html/Debuginfod-Settings.html
And the environment variables:
https://manpages.debian.org/testing/debuginfod/debuginfod.8.en.html#ENVIRONMENT_VARIABLES

kevinfrei replied:
@mysterymath that environment variable is used by the server, not the client.

In the interest of moving forward, I'd propose that this code is left as-is, and I'll add logging in libdebuginfod (separate from LLDB, and honestly, probably more useful) in a subsequent PR. I'll probably need this capability for my next steps, anyway (source code is next on my list)

mysterymath replied:
> @mysterymath that environment variable is used by the server, not the client.
Ah, I should have linked to the debuginfod-find man page instead; pardon my laziness. It is used by the client, it's just that the server also includes a client for federation purposes.

See https://sourceware.org/git/?p=elfutils.git;a=blob;f=debuginfod/debuginfod-client.c;h=dcf0503163fd73e126a2197d278f363fd6839810;hb=HEAD#l1881

kevinfrei replied:
I'm going to add better diagnostics to the llvm library instead, thus I didn't make a change to this in the updated diff...

walter-erquinigo commented on lldb/source/Symbol/LocateSymbolFile.cpp:
```diff
@@ -396,8 +398,22 @@ Symbols::LocateExecutableSymbolFile(const ModuleSpec &module_spec,
+    consumeError(result.takeError());
+  }
+  // Just return the empty FileSpec if nothing was found.
+  return dsym_bundle;
```
better return `{}`. It's more readable.

walter-erquinigo reviewed (commented):
pretty nice feature!

bulbazord reviewed (commented):
Out of curiosity, why did you choose the delimiter as ' ' instead of something like ';'?

clayborg reviewed (commented):
I would like to see a new setting added to lldb that should be the default way to enable the debuginfod support. Just adding something to `lldb/source/Target/TargetProperties.td` like:
```
  def DebuginfodURLs: Property<"debuginfod-urls", "Args">,
    Global,
    DefaultStringValue<"">,
    Desc<"A list of debuginfod URLs that will be linearly called to search for debug info.">;
```
Doesn't need to live in "target.*" though.

clayborg commented:
> I would like to see a new setting added to lldb that should be the default way to enable the debuginfod support. Just adding something to `lldb/source/Target/TargetProperties.td` like:
> 
> ```
>   def DebuginfodURLs: Property<"debuginfod-urls", "Args">,
>     DefaultStringValue<"">,
>     Desc<"A list of debuginfod URLs that will be linearly called to search for debug info.">;
> ```
> 
> Doesn't need to live in "target.*" though.

If this was added in target, then we could do:
```
(lldb) settings set target.debuginfod-urls http://url1 http://url2 http://url3
```

jimingham commented on lldb/source/Target/Target.cpp:
```diff
@@ -4892,6 +4894,21 @@ void TargetProperties::SetDebugUtilityExpression(bool debug) {
+  llvm::SmallVector<llvm::StringRef> dbginfod_urls;
+  std::transform(urls.begin(), urls.end(), dbginfod_urls.end(),
+                 [](const auto &obj) { return obj.ref(); });
+  llvm::setDefaultDebuginfodUrls(dbginfod_urls);
```
What happens here if I have two targets that want to use different URL's?

JDevlieghere replied:
Should this be a debugger (rather than a target) setting? In practice, is it likely to have multiple instances running or this more of a global resource (like dsymForUUID)?

kevinfrei replied:
I believe this should be more of a global resource, so moving the setting to Debugger makes sense.

jimingham replied:
This one is actually in kind of an awkward place in lldb.  

OTOH, if you have two different targets in lldb, they could be "one target remote debugging an Linux binary" and "one target debugging a local macOS binary".  So you would expect to be able to have different debuginfod sources for those very different binaries.  

However, because debug info & symbol parsing is slow, and a very common usage pattern for debug sessions is to  run the target multiple times where the main binary changes but everything else stays the same, lldb gets a good performance boost in this scenario by caching any "Modules" it has seen in a "global module cache".  

Also, when used in an IDE, you end up wanting one Debugger per target, not one Debugger and many targets, because the Debugger holds the command interpreter and you want each "debug session" to have a different console.  But in this case it's very common for all the Debugger sessions to be local to this machine, or to the same device, so to get the full benefits of the shared module cache it's held outside of the Debuggers.  So if a symbol-ingestion controlling setting was on the Target level, you could have two targets sharing the same Module, and if they have different debugInfod sources then it's a race to see which one gets queried, which is not so great.

So ideally you would like this to live in the Target because that's the level at which the decision of which source to choose is naturally made.  But because of this architectural oddity, that's a little hard to do internally.  Putting it as a global thing on Target will make something that's shared by all targets, but it gives a false impression that this will be target specific, when it currently can't be.

kevinfrei replied:
It's now a SymbolLocator plugin, so the setting is for the plugin, and follows that model of settings. I *think* this works the way you described.

JDevlieghere commented:
First off, thank you for working on this. `debuginfod` has been on my radar since support was added to LLVM and I was curious at which point someone was going to add support for it to LLDB. I wasn't super familiar with what exactly it provides, and in case others here aren't either, I found the following documentation very helpful: https://llvm.org/doxygen/Debuginfod_8h.html. 
TL;DR: it gives you a way to get sources, (symbol rich) executables and debug info files. 

The functionality `debuginfod` offers seems very similar to `dsymForUUID` (see https://lldb.llvm.org/use/symbols.html). We have series of functions such as `LocateExecutableObjectFile` and `LocateExecutableSymbolFile` in `LocateSymbolFile.h` that abstract over this. It seems appropriate that we use the same abstractions for `debuginfod` which avoids the special casing this patch introduces and a lot of things will "just work". 

The one caveat is that the current implementation of is really tied to a platform. All the `dsymForUUID` stuff is implemented in `LocateSymbolFileMacOSX.cpp`. I think we might need to move this to a Plugin model where we can have multiple implementations. The plugin name `SymbolVendor` is already taken so I would call this the `SymbolServer` plugin. Converting the current model to a plugin should be done in separate patch and myself or other Apple folks can make sure this doesn't break the dsymForUUID paths (we have some tests but not enough given how critical this is for us). With that done, it should be trivial to add a plugin implementation that calls into `debuginfod`

kevinfrei commented:
> First off, thank you for working on this. `debuginfod` has been on my radar since support was added to LLVM and I was curious at which point someone was going to add support for it to LLDB. I wasn't super familiar with what exactly it provides, and in case others here aren't either, I found the following documentation very helpful: https://llvm.org/doxygen/Debuginfod_8h.html. TL;DR: it gives you a way to get sources, (symbol rich) executables and debug info files.
> 
> The functionality `debuginfod` offers seems very similar to `dsymForUUID` (see https://lldb.llvm.org/use/symbols.html). We have series of functions such as `LocateExecutableObjectFile` and `LocateExecutableSymbolFile` in `LocateSymbolFile.h` that abstract over this. It seems appropriate that we use the same abstractions for `debuginfod` which avoids the special casing this patch introduces and a lot of things will "just work".
> 
> The one caveat is that the current implementation is really tied to a platform. All the `dsymForUUID` stuff is implemented in `LocateSymbolFileMacOSX.cpp`. I think we might need to move this to a Plugin model where we can have multiple implementations. The plugin name `SymbolVendor` is already taken so I would call this the `SymbolServer` plugin. Converting the current model to a plugin should be done in separate patch and myself or other Apple folks can make sure this doesn't break the dsymForUUID paths (we have some tests but not enough given how critical this is for us). With that done, it should be trivial to add a plugin implementation that calls into `debuginfod`

Yes, that specific kind of refactoring seemed like a good idea, but given that this is my first real foray into the LLDB space, I didn't want to bite off that much work to start with. Once I've got the full DEBUGINFOD capabilities plumbed, refactoring it out into a SymbolServer plug-in makes lots of sense.

You can also read about the full DebugInfoD protocol (it's not very complicated) [from the RedHat blog introducing it in 2019](https://developers.redhat.com/blog/2019/10/14/introducing-debuginfod-the-elfutils-debuginfo-server#).

kevinfrei commented on lldb/source/Symbol/CMakeLists.txt:
```diff
@@ -48,6 +48,7 @@ add_lldb_library(lldbSymbol NO_PLUGIN_DEPENDENCIES
     lldbHost
     lldbTarget
     lldbUtility
+    LLVMDebuginfod
```
`LINK_COMPONENTS` gets added to targets all over the place. The library is only used in this particular target, and any target that uses it, so this seemed like the minimal change necessary to make it work. I'm happy to put it in `LINK_COMPONENTS` if that's the right way to do it. Can anyone want to chime in who understands the build system more deeply than me? (I'm new to the community, so I have no idea who should make the call)

JDevlieghere commented:
> Yes, that specific kind of refactoring seemed like a good idea, but given that this is my first real foray into the LLDB space, I didn't want to bite off that much work to start with. 

I'd be happy to help with that and I'm sure @clayborg wouldn't mind providing guidance in this space. 

> Once I've got the full DEBUGINFOD capabilities plumbed, refactoring it out into a SymbolServer plug-in makes lots of sense.

That's fine if the new code can be more self contained. This patch is adding things like a `HTTPClient` to the debugger or a dependency on `Debuginfod.h` in Target. I don't think either of those things belong there. At the very least should be abstracted behind `LocateSymbol`. All its functions are static which makes that hard, which is why I suggested symbol server plugin. Maybe as an intermediate step we can start by turning that class into something that can be instantiated?

mysterymath commented on lldb/include/lldb/Target/Target.h:
```diff
@@ -258,6 +258,8 @@ class TargetProperties : public Properties {
 
   bool GetDebugUtilityExpression() const;
 
+  Args GetDebugInfoDURLs() const;
```
General naming nit:

`debuginfod` is all lower-case, and capitalized as if it were a single word: `Debuginfod`.

kevinfrei replied:
I *really* hate that, but that's clearly consistent with the rest of the code. I need to pronounce it as a single word, I think. "Dehbooginfawd"

mysterymath replied:
I tend to think of it as agglutanitive, like "blackbird."

kevinfrei commented:
> Out of curiosity, why did you choose the delimiter as ' ' instead of something like ';'?

Because that's how the environment variable works. It was less a choice and more 🤷

kevinfrei commented:
> > Yes, that specific kind of refactoring seemed like a good idea, but given that this is my first real foray into the LLDB space, I didn't want to bite off that much work to start with.
> 
> I'd be happy to help with that and I'm sure @clayborg wouldn't mind providing guidance in this space.
> 
> > Once I've got the full DEBUGINFOD capabilities plumbed, refactoring it out into a SymbolServer plug-in makes lots of sense.
> 
> That's fine if the new code can be more self contained. This patch is adding things like a `HTTPClient` to the debugger or a dependency on `Debuginfod.h` in Target. I don't think either of those things belong there. At the very least should be abstracted behind `LocateSymbol`. All its functions are static which makes that hard, which is why I suggested symbol server plugin. Maybe as an intermediate step we can start by turning that class into something that can be instantiated?

Are you asking me to create a SymbolServer class for *this* change, or do you want me to get this diff polished & landed, then create the abstraction before adding anything more? (Either is fine: I just can't quite tell what you're looking for right now)

I also agree: it made me someone uncomfortable adding those dependencies where they were added. Even just hiding them in a SymbolServer.h file would help.

kevinfrei committed:
All the non-contentious, obvious changes from PR feedback are in

mysterymath commented:
> > Out of curiosity, why did you choose the delimiter as ' ' instead of something like ';'?
> 
> Because that's how the environment variable works. It was less a choice and more 🤷
> 
> Also, ChatGPT tells me that URL's can include semicolons, so maybe it's not a great idea to use that as a delimiter? (I don't think Debuginfod server URL's can include semicolons, but I know they can't include spaces...)

I'd also recommend having a LLDB setting reuse the DEBUGINFOD_URLS syntax, just to avoid creating an unnecessary difference between the LLDB and GDB syntax.

JDevlieghere commented:
> Are you asking me to create a SymbolServer class for _this_ change, or do you want me to get this diff polished & landed, then create the abstraction before adding anything more? (Either is fine: I just can't quite tell what you're looking for right now)

I was suggesting the latter. 

As I was thinking about this a bit more and found some spare time, I gave the plugin conversion a try. I've created a PR (https://github.com/llvm/llvm-project/pull/71151) with the initial scaffolding and `LocateExecutableObjectFile` moved  into the plugin. Once that's complete, everything you're adding should fit perfectly in a new Debuginfod SymbolLocator plugin:

- The call to `llvm::HTTPClient::initialize()` can be done in the plugin instantiation.
- Plugins can have their own settings so we can set the URLs there. 
- Because of the way plugins work, they are tried in the order defined in the CMakeLists.txt, so if you make sure the `debuginfod` comes last, it'll be the fallback automatically.

JDevlieghere commented:
FWIW the "plugin-ification" has landed so this should be able to move forward as a plugin.

kevinfrei committed:
Moved to a plugin model

Commit:
Cleaned up leftovers from pre-plugin work

kevinfrei commented:
I updated the diff with the plugin-ified version of the work. It's *much* cleaner with no debugger-wide changes to speak of (Thanks for the set up for that, @JDevlieghere!). I did *not* add Debuginfod logging in the LLDB part of the code, as I intend to add diagnostic debugging to the llvm-debuginfod library instead, such that all users will benefit from it.

clayborg commented on lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp:
```diff
@@ -0,0 +1,154 @@
+      GetPluginNameStatic(), GetPluginDescriptionStatic(), CreateInstance,
+      LocateExecutableObjectFile, LocateExecutableSymbolFile,
+      DownloadObjectAndSymbolFile);
+  // There's a "safety" concern on this:
+  // Does plugin initialization occur while things are still single threaded?
```
This Initialize() function will get called when `lldb::SBDebugger::Initialize()` is called for the first time. `lldb::SBDebugger::Initialize()` and `lldb::SBDebugger::Terminate()` are ref counted, and only the first call will call `lldb::SBDebugger::Initialize()`. Since LLDB is a shared library, we shouldn't do any work unless clients ask for it. So it really depends on what program is loading LLDB. Command line lldb will init on the main thread right away, so will `lldb-vscode`/`lldb-dap` (VS Code IDE), so will `lldb-rpc-server` (Xcode), not sure about Android Studio or other IDEs

kevinfrei replied:
I'll update the comment, as I verified it was correct for normal LLDB. Is there somewhere we could document the requirement somehow for hosts? (or is there some mechanism for hosts that would allow this to be handled safely?)

kevinfrei replied:
It's part of the plugin system now, which works "as properly" as can be done in the world where library users might be doing crazy stuff.

clayborg commented on lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp:
```diff
@@ -0,0 +1,154 @@
+
+void SymbolLocatorDebuginfod::Terminate() {
+  PluginManager::UnregisterPlugin(CreateInstance);
+  // There's a "safety" concern on this:
+  // Does plugin termination occur while things are still single threaded?
```
Most probably won't be signle threaded. If this is a problem, you might leave this call out.

kevinfrei replied:
Nope. Dug in, looks okay.

clayborg commented on lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfodProperties.td:
```diff
@@ -0,0 +1,8 @@
+include "../../../../include/lldb/Core/PropertiesBase.td"
+
+let Definition = "symbollocatordebuginfod" in {
+  def URLs: Property<"urls", "String">,
+    Global,
+    DefaultStringValue<"">,
+    Desc<"A space-separated, ordered list of Debuginfod server URLs to query for symbols">;
+}
```
What does this look like on the command line? Other plug-in settings look like this:
```
plugin.dynamic-loader.darwin-kernel.load-kexts (boolean) = true
plugin.dynamic-loader.darwin-kernel.scan-type (enum) = fast-scan
plugin.jit-loader.gdb.enable (enum) = default
plugin.object-file.pe-coff.abi (enum) = default
plugin.object-file.pe-coff.module-abi (dictionary of enums) =
plugin.process.kdp-remote.packet-timeout (unsigned) = 5
plugin.process.gdb-remote.packet-timeout (unsigned) = 5
plugin.process.gdb-remote.target-definition-file (file) = 
plugin.process.gdb-remote.use-g-packet-for-reading (boolean) = false
plugin.process.gdb-remote.use-libraries-svr4 (boolean) = true
plugin.symbol-file.dwarf.ignore-file-indexes (boolean) = false
plugin.structured-data.darwin-log.auto-enable-options (string) = 
plugin.structured-data.darwin-log.enable-on-startup (boolean) = false
```
So hopefully this show up in `settings set plugin.symbolvendor.symbollocatordebuginfod.urls <url1> [<url2>]`

kevinfrei replied:
It looks like this now:
```
...
plugin.symbol-file.dwarf.ignore-file-indexes (boolean) = false
plugin.symbol-locator.debuginfod.server_urls (array of strings) =
  [0]: "https://debuginfod.centos.org/"
plugin.structured-data.darwin-log.auto-enable-options (string) =
...
```

clayborg commented on lldb/source/Target/Target.cpp:
```diff
@@ -4180,7 +4180,6 @@ TargetProperties::TargetProperties(Target *target)
         ePropertyInheritTCC, [this] { InheritTCCValueChangedCallback(); });
     m_collection_sp->SetValueChangedCallback(
         ePropertyDisableSTDIO, [this] { DisableSTDIOValueChangedCallback(); });
-
```
revert this

clayborg commented on llvm/include/llvm/Debuginfod/Debuginfod.h:
```diff
@@ -46,6 +46,10 @@ bool canUseDebuginfod();
 /// environment variable.
 SmallVector<StringRef> getDefaultDebuginfodUrls();
 
+/// Sets the list of debuginfod server URLs to query. This overrides the
+/// environment variable DEBUGINFOD_URLS.
```
Should this override the env vars? Or maybe add to them? Maybe we add a bool parameter to allow the user to choose like `bool append`?

kevinfrei replied:
Switched the setting to an array of strings setting, which addresses the usage issues.

clayborg commented on llvm/lib/Debuginfod/Debuginfod.cpp:
```diff
@@ -62,15 +66,25 @@ bool canUseDebuginfod() {
 }
 
+// Override the default debuginfod URL list.
+void setDefaultDebuginfodUrls(SmallVector<StringRef> URLs) {
+  DebuginfodUrls.clear();
```
```
void setDefaultDebuginfodUrls(SmallVector<StringRef> URLs, bool append) {
  if (!append)
    DebuginfodUrls.clear();
```

kevinfrei replied:
Switched the setting to an array of strings setting, which addresses the usage issues.

clayborg commented on lldb/source/Core/CoreProperties.td:
```diff
@@ -4,7 +4,7 @@ let Definition = "modulelist" in {
     Global,
     DefaultTrue,
-    Desc<"Control the use of external tools and repositories to locate symbol files. Directories listed in target.debug-file-search-paths and directory of the executable are always checked first for separate debug info files. Then depending on this setting: On macOS, Spotlight would be also used to locate a matching .dSYM bundle based on the UUID of the executable. On NetBSD, directory /usr/libda ...
+    Desc<"Control the use of external tools and repositories to locate symbol files. Directories listed in target.debug-file-search-paths and directory of the executable are always checked first for separate debug info files. Then depending on this setting: On macOS, Spotlight would be also used to locate a matching .dSYM bundle based on the UUID of the executable. On NetBSD, directory /usr/libda ...
```
We we want to mention the new setting users can set here as well for debuginfod?

kevinfrei replied:
Significantly changed the wording, here.

kevinfrei commented on lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp:
```diff
@@ -0,0 +1,154 @@
+#include <cstring>
+#include <optional>
+
+#include "Plugins/ObjectFile/wasm/ObjectFileWasm.h"
+#include "lldb/Core/Debugger.h"
+#include "lldb/Core/Module.h"
+#include "lldb/Core/ModuleList.h"
+#include "lldb/Core/ModuleSpec.h"
+#include "lldb/Core/PluginManager.h"
+#include "lldb/Core/Progress.h"
+#include "lldb/Core/Section.h"
+#include "lldb/Host/FileSystem.h"
+#include "lldb/Host/Host.h"
+#include "lldb/Symbol/ObjectFile.h"
+#include "lldb/Target/Target.h"
+#include "lldb/Utility/ArchSpec.h"
+#include "lldb/Utility/DataBuffer.h"
+#include "lldb/Utility/DataExtractor.h"
+#include "lldb/Utility/LLDBLog.h"
+#include "lldb/Utility/Log.h"
+#include "lldb/Utility/StreamString.h"
+#include "lldb/Utility/Timer.h"
+#include "lldb/Utility/UUID.h"
```
Whoops. You're right. Copy pasta is delicious...
I'll trim this down...a lot :/

Commit:
The server-urls property appears to be wired up correctly

kevinfrei commented:
Looks like I missed committing some staged edits. Hold on before reviewing...

bulbazord commented on lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp:
```diff
@@ -0,0 +1,142 @@
+}
+
+SymbolLocator *SymbolLocatorDebuginfod::CreateInstance() {
+  return new SymbolLocatorDebuginfod();
```
You're calling `llvm::canUseDebuginfod` in the methods to actually locate the debug info. Would it not make sense to call that when creating an instance? What would the value be in having a `SymbolLocatorDebuginfod` instance if `llvm::canUseDebuginfod` returns false?

kevinfrei replied:
canUseDebuginfod returns false if the Server URL list is empty. So if someone sets it after the plugin is created, it would have already failed.

bulbazord commented on llvm/lib/Debuginfod/Debuginfod.cpp:
```diff
@@ -62,15 +66,25 @@ bool canUseDebuginfod() {
+  if (!DebuginfodUrlsSet) {
+    // Only read from the environment variable if the user hasn't already
+    // set the value
+    const char *DebuginfodUrlsEnv = std::getenv("DEBUGINFOD_URLS");
+    if (DebuginfodUrlsEnv != nullptr) {
+      StringRef(DebuginfodUrlsEnv).split(DebuginfodUrls, " ", -1, false);
+    }
```
```suggestion
    if (const char *DebuginfodUrlsEnv = std::getenv("DEBUGINFOD_URLS"))
      StringRef(DebuginfodUrlsEnv).split(DebuginfodUrls, " ", -1, false);
```

kevinfrei committed:
Update llvm/lib/Debuginfod/Debuginfod.cpp

Co-authored-by: Alex Langford <nirvashtzero@gmail.com>

clayborg commented on llvm/lib/Debuginfod/Debuginfod.cpp:
```diff
@@ -62,15 +66,23 @@ bool canUseDebuginfod() {
+  if (!DebuginfodUrlsSet) {
+    // Only read from the environment variable if the user hasn't already
+    // set the value
+    if (const char *DebuginfodUrlsEnv = std::getenv("DEBUGINFOD_URLS"))
+      StringRef(DebuginfodUrlsEnv).split(DebuginfodUrls, " ", -1, false);
+    DebuginfodUrlsSet = true;
```
If we use std::optional, this can be:
```
DebuginfodUrls = SmallVector<StringRef>();
if (const char *DebuginfodUrlsEnv = std::getenv("DEBUGINFOD_URLS"))
  StringRef(DebuginfodUrlsEnv).split(DebuginfodUrls.value(), " ", -1, false);
```

kevinfrei replied:
Much better

clayborg commented on llvm/lib/Debuginfod/Debuginfod.cpp:
```diff
@@ -62,15 +66,23 @@ bool canUseDebuginfod() {
 
+// Set the default debuginfod URL list, override the environment variable
+void setDefaultDebuginfodUrls(SmallVector<StringRef> URLs) {
+  DebuginfodUrls.clear();
+  DebuginfodUrls.insert(DebuginfodUrls.begin(), URLs.begin(), URLs.end());
+  DebuginfodUrlsSet = true;
```
If we use std::optional this becomes:
```
DebuginfodUrls = URLs;
```
And the std::optional will have a new value and the value will be set.

kevinfrei replied:
So much cleaner!

clayborg commented on llvm/include/llvm/Debuginfod/Debuginfod.h:
```diff
@@ -46,6 +46,10 @@ bool canUseDebuginfod();
 
+/// Sets the list of debuginfod server URLs to query. This overrides the
+/// environment variable DEBUGINFOD_URLS.
+void setDefaultDebuginfodUrls(SmallVector<StringRef> URLs);
```
pass in a const reference to avoid copying on calling this?:
```
void setDefaultDebuginfodUrls(const SmallVector<StringRef> &URLs);
```

kevinfrei replied:
Done.

clayborg reviewed (commented):
Looks good, a few suggestions in `llvm/lib/Debuginfod/Debuginfod.cpp`. I will let Jonas comment if the settings are done the right way in the plug-in manager. Plug-in code looks fine.

clayborg commented on lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp:
```diff
@@ -0,0 +1,142 @@
+
+std::optional<FileSpec> SymbolLocatorDebuginfod::LocateExecutableSymbolFile(
+    const ModuleSpec &module_spec, const FileSpecList &default_search_paths) {
+  const UUID &module_uuid = module_spec.GetUUID();
+  if (module_uuid.IsValid() && llvm::canUseDebuginfod()) {
+    llvm::object::BuildID build_id(module_uuid.GetBytes());
+    llvm::Expected<std::string> result =
+        llvm::getCachedOrDownloadDebuginfo(build_id);
+    if (result)
+      return FileSpec(*result);
+    // An error here should be logged as a failure in the Debuginfod library,
+    // so just consume it here
+    consumeError(result.takeError());
+  }
```
This implementation is exactly the same as `SymbolLocatorDebuginfod::LocateExecutableObjectFile(...)` right? We should just call that directly from here if so?

kevinfrei replied:
I factored it into a helper to reduce copy-pasta

JDevlieghere reviewed (approved):
I was out last week. Thanks, this LGTM!

Commit:
Fixed memory ownership issues, and @clayborg's feedback

kevinfrei commented on llvm/lib/Debuginfod/Debuginfod.cpp:
```diff
@@ -47,6 +47,10 @@ namespace llvm {
 
 using llvm::object::BuildIDRef;
 
+SmallVector<StringRef> DebuginfodUrls;
```
Done

clayborg commented:
LGTM.

DavidSpickett commented:
If this code is calling a library - which itself is already tested, that's fine but any smoke test we can do would be good.

E.g. set a bogus URL and expect that it fails to use it (assuming it does so visibly). Just to check the integration.

Commit:
Fixed the autoformatting stuff

Commit:
Added a release note item for DEDBUGINFOD

clayborg reviewed (changes requested):
We need to respect the setting `symbols.enable-external-lookup` and not do anything with debuginfod if this is set to false.

You can get this value by calling:
```
ModuleList::GetGlobalModuleListProperties().GetEnableExternalLookup()
```
Otherwise I fear that the linux buildbots will start failing if we get extra system library symbols if debuginfod is enabled via the environment variable.

Commit:
Added code to respect the symbols.enable-external-lookup setting

kevinfrei commented:
> We need to respect the setting `symbols.enable-external-lookup` and not do anything with debuginfod if this is set to false.
> 
> You can get this value by calling:
> 
> ```
> ModuleList::GetGlobalModuleListProperties().GetEnableExternalLookup()
> ```
> 
> Otherwise I fear that the linux buildbots will start failing if we get extra system library symbols if debuginfod is enabled via the environment variable.

Fixed that up, rebased and push -f 'ed

nico commented:
This breaks targeting macOS versions older than 10.12 for unnecessary reason:

```
../../llvm/lib/Debuginfod/Debuginfod.cpp:54:6: error: 'shared_mutex' is unavailable: introduced in macOS 10.12
std::shared_mutex UrlsMutex;
     ^
```

You should be able to use llvm::sys::RWMutex so please switch to that. See https://reviews.llvm.org/D138423 for an example.

kevinfrei commented:
> This breaks targeting macOS versions older than 10.12 for unnecessary reasons:
> 
> ```
> ../../llvm/lib/Debuginfod/Debuginfod.cpp:54:6: error: 'shared_mutex' is unavailable: introduced in macOS 10.12
> std::shared_mutex UrlsMutex;
>      ^
> ```
> 
> You should be able to use llvm::sys::RWMutex so please switch to that. See https://reviews.llvm.org/D138423 for an example.

Noooooooooooo!

I'll get this fixed up ASAP. Should I open an issue, or just put up a new PR? (Thanks for the example!)

mgorny commented:
This change broke building against LLVM dylib:

```
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: cannot find -lLLVMDebuginfod: No such file or directory
collect2: error: ld returned 1 exit status
samu: subcommand failed
```

Full build log: 
[dev-util:lldb-18.0.0_pre20231206:20231206-163149.log](https://github.com/llvm/llvm-project/files/13588891/dev-util.lldb-18.0.0_pre20231206.20231206-163149.log)

kevinfrei commented:
> This change broke building against LLVM dylib:
> 
> ```
> /usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: cannot find -lLLVMDebuginfod: No such file or directory
> collect2: error: ld returned 1 exit status
> samu: subcommand failed
> ```
> 
> Full build log: [dev-util:lldb-18.0.0_pre20231206:20231206-163149.log](https://github.com/llvm/llvm-project/files/13588891/dev-util.lldb-18.0.0_pre20231206.20231206-163149.log)

Sounds like I get to dig in deeper to LLVM's build, now. Hooray 🤣

kevinfrei commented:
> This change broke building against LLVM dylib:
> 
> ```
> /usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: cannot find -lLLVMDebuginfod: No such file or directory
> collect2: error: ld returned 1 exit status
> samu: subcommand failed
> ```
> 
> Full build log: [dev-util:lldb-18.0.0_pre20231206:20231206-163149.log](https://github.com/llvm/llvm-project/files/13588891/dev-util.lldb-18.0.0_pre20231206.20231206-163149.log)

@mgorny Can you please point me at where to find all the configuration for those builds? I'm stuck on a weirdo CentOS build (or MacOS) and can't manage to get my CMake configuration stuff to let me build a dylib/so. I'll get on this first thing tomorrow.

kevinfrei commented:
> > This change broke building against LLVM dylib:
> > ```
> > /usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: cannot find -lLLVMDebuginfod: No such file or directory
> > collect2: error: ld returned 1 exit status
> > samu: subcommand failed
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Full build log: [dev-util:lldb-18.0.0_pre20231206:20231206-163149.log](https://github.com/llvm/llvm-project/files/13588891/dev-util.lldb-18.0.0_pre20231206.20231206-163149.log)
> 
> @mgorny Can you please point me at where to find all the configuration for those builds? I'm stuck on a weirdo CentOS build (or MacOS) and can't manage to get my CMake configuration stuff to let me build a dylib/so. I'll get on this first thing tomorrow.


> This change broke building against LLVM dylib:
> 
> ```
> /usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: cannot find -lLLVMDebuginfod: No such file or directory
> collect2: error: ld returned 1 exit status
> samu: subcommand failed
> ```
> 
> Full build log: [dev-util:lldb-18.0.0_pre20231206:20231206-163149.log](https://github.com/llvm/llvm-project/files/13588891/dev-util.lldb-18.0.0_pre20231206.20231206-163149.log)

@mgorny I've burned 3 hours this morning trying to get the visible configuration flags from that log file to function _before_ my patch on (my weirdo CentOS 9) Linux box, or cause problems _after_ my patch on macOS, and failed on both fronts. I'm stuck until you can get me the contents of `gentoo_common_config.cmake` (or tell me where to find it...)

clayborg commented:
Sounds like some configs don't create `LLVMDebuginfod` target possibly because it isn't enabled. Is there a cmake option to disable `LLVMDebuginfod` or some magic in the Cmake config file for this target?

kevinfrei commented:
> Sounds like some configs don't create `LLVMDebuginfod` target possibly because it isn't enabled. Is there a cmake option to disable `LLVMDebuginfod` or some magic in the Cmake config file for this target?

Nothing that I could find. The library is written such that if the configuration doesn't include either libcurl or httplib then all the functions just fail to do anything, so from what I can tell there'd never be a reason to disable it.

mgorny commented:
I'm sorry for not getting back to you earlier. The log relevant to building LLVM itself is this (22M uncompressed):

[sys-devel:llvm-18.0.0_pre20231206:20231206-072428.log.gz](https://github.com/llvm/llvm-project/files/13627566/sys-devel.llvm-18.0.0_pre20231206.20231206-072428.log.gz)

I think the most relevant part is:

```
-DLLVM_DISTRIBUTION_COMPONENTS="LLVM;LTO;Remarks;llvm-config;cmake-exports;llvm-headers;LLVMDemangle;LLVMSupport;LLVMTableGen;llvm_gtest;llvm_gtest_main;LLVMTestingAnnotations;LLVMTestingSupport;llvm-tblgen;FileCheck;llvm-PerfectShuffle;count;not;yaml-bench;UnicodeNameMappingGenerator;bugpoint;dsymutil;llc;lli;lli-child-target;llvm-addr2line;llvm-ar;llvm-as;llvm-bcanalyzer;llvm-bitcode-strip;llvm-c-test;llvm-cat;llvm-cfi-verify;llvm-config;llvm-cov;llvm-cvtres;llvm-cxxdump;llvm-cxxfilt;llvm-cxxmap;llvm-debuginfo-analyzer;llvm-debuginfod-find;llvm-diff;llvm-dis;llvm-dlltool;llvm-dwarfdump;llvm-dwarfutil;llvm-dwp;llvm-exegesis;llvm-extract;llvm-gsymutil;llvm-ifs;llvm-install-name-tool;llvm-jitlink;llvm-jitlink-executor;llvm-lib;llvm-libtool-darwin;llvm-link;llvm-lipo;llvm-lto;llvm-lto2;llvm-mc;llvm-mca;llvm-ml;llvm-modextract;llvm-mt;llvm-nm;llvm-objcopy;llvm-objdump;llvm-opt-report;llvm-otool;llvm-pdbutil;llvm-profdata;llvm-profgen;llvm-ranlib;llvm-rc;llvm-readelf;llvm-readobj;llvm-readtapi;llvm-reduce;llvm-remarkutil;llvm-rtdyld;llvm-sim;llvm-size;llvm-split;llvm-stress;llvm-strings;llvm-strip;llvm-symbolizer;llvm-tli-checker;llvm-undname;llvm-windres;llvm-xray;obj2yaml;opt;sancov;sanstats;split-file;verify-uselistorder;yaml2obj;opt-viewer;docs-dsymutil-man;docs-llvm-dwarfdump-man;docs-llvm-man;docs-llvm-html;LLVMgold;llvm-debuginfod"
```

We don't install `LLVMDebuginfod.a` at all right now. I suppose I could try doing that but last we tried to do something similar, all hell broke loose because of its dependency on other libraries from LLVM. It's a mess, I'm afraid.

At this point, I think the only reasonable solution here is to include `LLVMDebuginfod` in `libLLVM`.

mgorny commented:
I've also confirmed that installing `LLVMDebuginfod` target won't help since it references other internal targets that are created at build-time and not included as part of the installed CMake configuration:

```
CMake Error at /usr/lib/llvm/18/lib64/cmake/llvm/LLVMExports.cmake:90 (set_target_properties):
  The link interface of target "LLVMDebuginfod" contains:

    httplib::httplib

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  /usr/lib/llvm/18/lib64/cmake/llvm/LLVMConfig.cmake:355 (include)
  cmake/modules/LLDBStandalone.cmake:9 (find_package)
  CMakeLists.txt:34 (include)
```

I suspect that even if we hadn't used the dylib at all and installed all static libraries, it would be impossible to build LLDB against this install of LLVM as it wouldn't know about `httplib::httplib` — that is, unless LLVM was built without httplib support.

kevinfrei commented:
Is there some documentation that both me & google are missing somewhere? I'm trying to read between the lines of the scenario that's broken, with nothing but a log (that refers to a bunch of pretty important configuration files from somewhere I can't find) with a repo configuration that, again, I can't figure out. You say that we can't "build against LLMV as a dylib". Is this trying to build LLDB stand-alone, without a full LLVM repo around it? I'm also assuming you're actually talking about an SO and not a dylib, since the build is on Linux, not macOS. I really want to help, but I'm recently new to this space, so I'm missing a bunch of tribal knowledge.

Alternatively, if you have a repro, and can fix it by adding LLVMDebuginfod to libLLVM, that seems reasonable to me, given the design of LLVMDebuginfod (small, falls back to no functionality without libcurl, etc...) I've burned all day today trying to get a build & repro going, and have failed (my build is quite complicated). I'll try this on a personal machine tomorrow which should be simpler (though a couple orders of magnitude slower) than my work behemoth...

mgorny commented:
"Dylib" is LLVM-speech for `libLLVM.so` ("shared libraries" refer to individual split libraries, while "dylib" refers to the monolith), as in `-DLLVM_BUILD_LLVM_DYLIB=ON -DLLVM_LINK_LLVM_DYLIB=ON`. And yes, that + standalone build with full *source* repository but linking against LLVM+Clang installed with `install-distribution` target.

And yeah, if you make a patch to add LLVMDebuginfod to libLLVM, I can test that here.

mgorny commented:
Did you `ninja install-distribution` rather than regular `install`? Did it pick the right installation? Could you attach the build logs?

kevinfrei commented:
Ran the first script, sudo nina install-distribution, then ran the second script.

I don't see the build depending on what's been installed, as it looks like setting LLVM_DIR makes the .so dependency on the binary sitting in the rel-llvm/lib directory instead.

[build.log.gz](https://github.com/llvm/llvm-project/files/13653581/build.log.gz)

mgorny commented:
I'm afraid that log's non-verbose (i.e. missing `-v`), so it doesn't tell me which libraries are being linked. Also, I don't think I'll be able to tell much without logs for all of cmake, build and install of both LLVM and LLDB.

kevinfrei commented:
I'll try to get this together tonight or tomorrow morning.

kevinfrei commented:
doBuild.sh:
```
#!/bin/sh
./llvm-config.sh
cd rel-llvm
ninja -v
sudo ninja -v install-distribution
cd ..
./lldb-config.sh
cd rel-lldb
ninja -v
```
llvm-config.sh:
```
#!/bin/sh
cmake -G Ninja \
  -B rel-llvm \
  -S llvm/llvm \
  -DCMAKE_BUILD_TYPE=Release \
  -DLLVM_ENABLE_PROJECTS=clang \
  -DCMAKE_INSTALL_PREFIX=/usr \
  -DLLVM_APPEND_VC_REV=OFF \
  -DCMAKE_INSTALL_PREFIX=/usr/lib/llvm/18 \
  -DBUILD_SHARED_LIBS=OFF \
  -DLLVM_BUILD_LLVM_DYLIB=ON \
  -DLLVM_LINK_LLVM_DYLIB=ON \
  -DLLVM_TARGETS_TO_BUILD="AArch64;X86" \
  -DLLVM_INCLUDE_BENCHMARKS=OFF \
  -DLLVM_ENABLE_ZSTD=yes \
  -DLLVM_ENABLE_CURL=yes \
  -DLLVM_HOST_TRIPLE=i686-pc-linux-gnu \
  -DLLVM_DISTRIBUTION_COMPONENTS="LLVM;LTO;Remarks;llvm-config;cmake-exports;llvm-headers;LLVMDemangle;LLVMSupport;LLVMTableGen;llvm-tblgen;dsymutil;llc;lli;llvm-addr2line;llvm-ar;llvm-as;llvm-bcanalyzer;llvm-bitcode-strip;llvm-c-test;llvm-cat;llvm-cfi-verify;llvm-config;llvm-cov;llvm-cvtres;llvm-cxxdump;llvm-cxxfilt;llvm-cxxmap;llvm-debuginfo-analyzer;llvm-debuginfod-find;llvm-diff;llvm-dis;llvm-dlltool;llvm-dwarfdump;llvm-dwarfutil;llvm-dwp;llvm-exegesis;llvm-extract;llvm-gsymutil;llvm-ifs;llvm-install-name-tool;llvm-jitlink;llvm-lib;llvm-libtool-darwin;llvm-link;llvm-lipo;llvm-lto;llvm-lto2;llvm-mc;llvm-mca;llvm-ml;llvm-modextract;llvm-mt;llvm-nm;llvm-objcopy;llvm-objdump;llvm-opt-report;llvm-otool;llvm-pdbutil;llvm-profdata;llvm-profgen;llvm-ranlib;llvm-rc;llvm-readelf;llvm-readobj;llvm-readtapi;llvm-reduce;llvm-remarkutil;llvm-rtdyld;llvm-sim;llvm-size;llvm-split;llvm-stress;llvm-strings;llvm-strip;llvm-symbolizer;llvm-tli-checker;llvm-undname;llvm-windres;llvm-xray;opt;sancov;sanstats;verify-uselistorder;opt-viewer;llvm-debuginfod"
```
lldb-config.sh:
```
#!/bin/sh
cmake -G Ninja \
  -B rel-lldb \
  -S llvm/lldb \
  -DCMAKE_BUILD_TYPE=Release \
  -DLLVM_TARGETS_TO_BUILD="AArch64;X86" \
  -DCMAKE_INSTALL_PREFIX=/usr \
  -DLLVM_ENABLE_LIBEDIT=yes \
  -DLLVM_ENABLE_TERMINFO=yes \
  -DLLVM_ENABLE_LIBXML2=yes \
  -DLLVM_ENABLE_EH=ON \
  -DLLVM_ENABLE_RTTI=ON \
  -DLLVM_ENABLE_CURL=yes \
  -DLLVM_HOST_TRIPLE=i686-pc-linux-gnu \
  -DPython3_EXECUTABLE=/usr/bin/python3 \
  -DLLVM_DIR=/home/freik/src/rel-llvm/lib/cmake/llvm
```
And, here's the output of that whole mess: 
[hopefully-enough.log.gz](https://github.com/llvm/llvm-project/files/13681262/hopefully-enough.log.gz)

mgorny commented:
Ping.

kevinfrei commented:
Sorry: nice long holiday break. So calling that library a "left over from a previous build" confuses me. What's the point of the `-DLLVM_DIR=/home/freik/src/rel-llvm/lib/cmake/llvm` argument to the second cmake config if not to instruct the build where to find stuff from the full LLVM build? If I delete that library, I can, indeed, reproduce the issue [now working on it] but that also leads me to believe that my repro still isn't actually doing what your configuration is. Because of that, I have no confidence that what I do to fix the issue is going to *actually* fix it 😞

mgorny commented:
Ah, sorry, I didn't notice that. We're not passing `LLVM_DIR` at all. LLDB is able to find installed LLVM & Clang via `PATH`, and that's all it needed.

kevinfrei commented:
I've returned to this after getting some other work up for a PR, and I'm stuck again. If I remove LLVM_DIR, the thing doesn't get anywhere. It explicitly asks for LLVM_DIR, if I work through that, then it asks for Clang_DIR. I'm getting frustrated because it seems like the configuration you're stuck on isn't supported (anymore: I'm guessing it's a holdover from before the monorepo). I just can't find any documentation for building the way you're describing. Everything says to use LLVM_DIR. So, if I set LLVM_DIR, it (correctly) links against libDebuginfod.a just fine. And if I don't have LLVM_DIR set, there'a whole lot of stuff that fails to work properly.

Is there some docker container I can spin up with your configuration already setup or something? I'm completely stuck again.

mgorny commented:
`LLVM_DIR` and `Clang_DIR` are merely hints to `find_package()`. If you need to pass them, it simply means you haven't set `PATH` correctly and CMake can't find installed LLVM/Clang. However, that shouldn't make any difference as long as you provide paths to **installed** LLVM/Clang, and not to the source/build directory. That said, you may actually need to remove the build directory because I don't know if things don't leak through.

No, I don't have a "ready" configuration. You could start off `gentoo/stage3` and install LLVM/Clang/LLDB live ebuilds to reproduce but that's a lot of effort, and I don't think it's worth your time.

Perhaps I should just send the "obvious" fix without bothering you. I can reproduce it anytime, and I doubt there's any other fix possible than adding `LLVMDebuginfod` library to `libLLVM`. I just need a lot of energy to start contributing to LLVM because it's painful.

kevinfrei commented:
Sure. That's probably the solution. I just can't get a reliable repro, so I can only validate that it doesn't break the mainstream monorepo build. I'm happy to shepherd any diff you've indicated can tell me it's fixed on your configuration.

I truly appreciate the different worlds, here. You toss out "install a particular build of gentoo, which you've probably never done" while complaining about having to clone a repo I have on 5 different machines 🤣

mgorny commented:
I'm sorry for the complexity. I've finally gotten around to figuring it out, and it turns out it's a problem with CMake files: #79305 fixes it for me.
